routers/web/repo/blame.go

```diff
@@ -164,6 +178,17 @@ func RefBlame(ctx *context.Context) {
 			return
 		}
 
+		// Get previous closest commit sha from the commit
+		previousCommit, err := commit.Parent(0)
+		if err == nil {
+			// Verify the commit
+			if haz, _ := commit.HasPreviousCommit(previousCommit.ID); haz {
+				if haz1, _ := previousCommit.HasFile(ctx.Repo.TreePath); haz1 {
+					previousCommits[commit.ID.String()] = previousCommit.ID.String()
+				}
+			}
+		}
+
```

Comment:
Argh!

commit.HasPreviousCommit is not free and is potentially computationally very expensive. It involves a call to `git merge-base --is-ancestor`. The results of this need to be cached at least within the loop. 

Further I don't think:

`haz, _ := commit.HasPreviousCommit(previousCommit.ID)` can ever be false as it is a 0-level parent by line 182: `previousCommit, err := commit.Parent(0)`.

`haz1, _ := previousCommit.HasFile(ctx.Repo.TreePath); haz1` - is also quite expensive and should be cached.

Reply:
Oh ok. Is there a common cache facility to use for this? Or should I just do an adhoc `map[string]bool` cache?

Reply:
probably just use a just out-of-loop cache for the moment. In this case I don't think there's likely to be benefit pushing up to `modules/cache` or using an `*lru.TwoQueue`/`*lru.Cache`  directly.

We should consider whether to cache the results of the internal `git.Commit.GetFiles` in modules/cache.Cache but at present the default cache is unlimited unless we merge my PR.

Reply:
```diff
diff --git a/routers/web/repo/blame.go b/routers/web/repo/blame.go
index 9efb868c8..79612f0d5 100644
--- a/routers/web/repo/blame.go
+++ b/routers/web/repo/blame.go
@@ -161,6 +161,8 @@ func RefBlame(ctx *context.Context) {
 	commitNames := make(map[string]models.UserCommit)
 	previousCommits := make(map[string]string)
 	commits := list.New()
+	commitCache := map[string]*git.Commit{}
+	commitCache[commitID] = commit
 
 	for _, part := range blameParts {
 		sha := part.Sha
@@ -168,21 +170,30 @@ func RefBlame(ctx *context.Context) {
 			continue
 		}
 
-		commit, err := ctx.Repo.GitRepo.GetCommit(sha)
-		if err != nil {
-			if git.IsErrNotExist(err) {
-				ctx.NotFound("Repo.GitRepo.GetCommit", err)
-			} else {
-				ctx.ServerError("Repo.GitRepo.GetCommit", err)
+		commit, ok := commitCache[sha]
+		if !ok {
+			commit, err = ctx.Repo.GitRepo.GetCommit(sha)
+			if err != nil {
+				if git.IsErrNotExist(err) {
+					ctx.NotFound("Repo.GitRepo.GetCommit", err)
+				} else {
+					ctx.ServerError("Repo.GitRepo.GetCommit", err)
+				}
+				return
 			}
-			return
+			commitCache[sha] = commit
 		}
 
-		// Get previous closest commit sha from the commit
-		previousCommit, err := commit.Parent(0)
-		if err == nil {
-			// Verify the commit
-			if haz, _ := commit.HasPreviousCommit(previousCommit.ID); haz {
+		if len(commit.Parents) > 0 {
+			psha := commit.Parents[0]
+			previousCommit, ok := commitCache[psha.String()]
+			if !ok {
+				previousCommit, _ = commit.Parent(0)
+				if previousCommit != nil {
+					commitCache[psha.String()] = previousCommit
+				}
+			}
+			if previousCommit != nil {
 				if haz1, _ := previousCommit.HasFile(ctx.Repo.TreePath); haz1 {
 					previousCommits[commit.ID.String()] = previousCommit.ID.String()
 				}
```

Reply:
thanks, done